src/System.IO.FileSystem/src/System/IO/WinRTFileSystem.cs

```diff
@@ -227,14 +228,121 @@ public override bool DirectoryExists(string fullPath)
 
         public override IEnumerable<string> EnumeratePaths(string fullPath, string searchPattern, SearchOption searchOption, SearchTarget searchTarget)
         {
-            // Fill in implementation
-            return new string[0];
+            IReadOnlyList<IStorageItem> storageFiles = SynchronousResultOf(EnumerateFileQuery(fullPath, searchPattern, searchOption, searchTarget));
+            return TranslatePaths(storageFiles);
         }
 
         public override IEnumerable<FileSystemInfo> EnumerateFileSystemInfos(string fullPath, string searchPattern, SearchOption searchOption, SearchTarget searchTarget)
         {
-            // Fill in implementation
-            return new FileSystemInfo[0];
+            IReadOnlyList<IStorageItem> storageFiles = SynchronousResultOf(EnumerateFileQuery(fullPath, searchPattern, searchOption, searchTarget));
+            return TranslateFileSystemInfos(storageFiles);
+        }
+
+        /// <summary>
+        /// Translates IStorageItems into FileSystemInfos
+        /// </summary>
+        private static IEnumerable<FileSystemInfo> TranslateFileSystemInfos(IReadOnlyList<IStorageItem> storageFiles)
+        {
+            // Translate the StorageFile results into FileSystemInfos
+            int count = storageFiles.Count;
+            for (int i = 0; i < count; i++)
+            {
+                if (storageFiles[i].IsOfType(StorageItemTypes.Folder))
+                {
+                    yield return new DirectoryInfo(storageFiles[i].Path);
+                }
+                else // If it is neither a File nor folder then we treat it as a File.
+                {
+                    yield return new FileInfo(storageFiles[i].Path);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Translates IStorageItems into string paths
+        /// </summary>
+        private static IEnumerable<string> TranslatePaths(IReadOnlyList<IStorageItem> storageFiles)
+        {
+            // Yield the paths of the StorageItems
+            int count = storageFiles.Count;
+            for (int i = 0; i < count; i++)
+            {
+                yield return storageFiles[i].Path;
+            }
+        }
+
+        private async static Task<IReadOnlyList<IStorageItem>> EnumerateFileQuery(string path, string searchPattern, SearchOption searchOption, SearchTarget searchTarget)
+        {
+            // Get a StorageFolder for "path"
+            string fullPath = Path.GetFullPath(path);
+            StorageFolder folder;
+            try
+            {
+                folder = await StorageFolder.GetFolderFromPathAsync(fullPath).TranslateWinRTTask(fullPath, isDirectory: true);
```

Comment:
> So  GetFolderFromPath.TranslateWinRtTask.ConfigureAwait ?

Yes.  Essentially anywhere in the library we're awaiting something, we should await it with ```.ConfigureAwait(false)```.

> We do not do that for any other awaits in this file that use TranslateWinRTTask. I take it we should be?

So, it depends.  The safest answer is "yes, use it everywhere in the library we await something".  There's a more nuanced answer as well, which is that if we know we've been queued to a thread pool thread (e.g. we just did Task.Run and we're inside the delegate or something like that), you could leave of the ConfigureAwait.  ```ConfigureAwait(false)``` basically overrides the default await behavior that looks for the current SynchronizationContext/TaskScheduler and schedules the continuation back to them; instead, with ConfigureAwait(false), it behaves as if both of those returned null/the default.  This avoids unnecessarily scheduling back to the current context, e.g. the UI thread if that's where the method is called from, and that has two benefits.  One, there's a perf win in avoiding that hop.  Two, if someone blocks waiting for the XxAsync method to return, and that Async method internally won't complete until some continuation it's scheduled executes, and if that continuation can only run on the very thread that's blocking waiting for the async operation to complete, you've now deadlocked... using ConfigureAwait helps avoid that.